File: javascript/testSalkin.js
const ConstrainedGA = require('./constrainedGA');

// Section 3.1: SALKIN Problem Setup
// Maximize: 3x1 + x2 + 2x3 + x4 - x5 -> Inverted to Minimize for the GA engine
const salkinObjective = (x) => -(3 * x[0] + x[1] + 2 * x[2] + x[3] - x[4]);

// Re-formulating g_i(x) <= Constant into g_i(x) - Constant <= 0
const g1 = (x) => (25 * x[0] - 40 * x[1] + 16 * x[2] + 21 * x[3] + x[4]) - 300;
const g2 = (x) => (x[0] + 20 * x[1] - 50 * x[2] + x[3] - x[4]) - 200;
const g3 = (x) => (60 * x[0] + x[1] - x[2] + 2 * x[3] + x[4]) - 600;
const g4 = (x) => (-7 * x[0] + 4 * x[1] + 15 * x[2] - x[3] + 65 * x[4]) - 700;

// Variable Bounds from paper: 5 dimensions
const salkinBounds = [
    [1.0, 4.0],     // x1
    [80.0, 88.0],   // x2
    [30.0, 35.0],   // x3
    [145.0, 150.0], // x4
    [0.0, 2.0]      // x5
];

const runsCount = 100;
const executionValues = [];
const executionFevals = [];

console.log(`Starting validation: Running SALKIN optimization ${runsCount} times...`);

for (let run = 1; run <= runsCount; run++) {
    const solver = new ConstrainedGA({
        objFunc: salkinObjective,
        inequalityConstraints: [g1, g2, g3, g4],
        bounds: salkinBounds,
        K: 200,       // K = 200
        iterMax: 200, // ITERMAX = 200
        lam: 1000.0   // \lambda = 10^3
    });

    const runResult = solver.solve();
    
    // Convert back to maximization scale by multiplying by -1
    executionValues.push(-runResult.value);
    executionFevals.push(runResult.fevalsSpent);
    
    if (run % 20 === 0) {
        console.log(`  > Completed ${run}/${runsCount} runs...`);
    }
}

// Compute statistics
const avgFevals = executionFevals.reduce((a, b) => a + b, 0) / runsCount;
const bestValue = Math.max(...executionValues); // Looking for max value now
const meanValue = executionValues.reduce((a, b) => a + b, 0) / runsCount;
const variance = executionValues.reduce((sum, val) => sum + Math.pow(val - meanValue, 2), 0) / runsCount;
const stdDeviation = Math.sqrt(variance);

// Benchmark Table Output
console.log("\n=======================================================");
console.log("           SALKIN PROBLEM BENCHMARK RESULTS            ");
console.log("=======================================================");
console.log(`Metric       |  Paper Table 2 Data  | Your Local Run JS`);
console.log("-------------------------------------------------------");
console.log(`FEVALS       |  7244               | ${Math.round(avgFevals)}`);
console.log(`BEST         |  320.0000           | ${bestValue.toFixed(4)}`);
console.log(`MEAN         |  320.0000           | ${meanValue.toFixed(4)}`);
console.log(`STD          |  0.0                | ${stdDeviation.toExponential(2)}`);
console.log("=======================================================");

File: matlab/testSalkin.m
% SALKIN Problem Test

clear; clc;

% Define objective function
salkinObjective = @(x) -5*x(1) - 5*x(2) - 5*x(3) - 5*x(4) + 100;

% Define constraints
g1 = @(x) 2*x(1) + 2*x(2) + x(3) + x(4) - 10;
g2 = @(x) -2*x(1) - 2*x(2) - x(3) - x(4) + 10;

% Variable bounds
salkinBounds = [0 1; 0 1; 0 1; 0 1];

% Number of runs
runsCount = 100;
executionValues = zeros(runsCount, 1);
executionFevals = zeros(runsCount, 1);

fprintf('Starting validation: Running SALKIN optimization %d times...\n', runsCount);

for run = 1:runsCount
    solver = ConstrainedGA('ObjectiveFunction', salkinObjective, ...
                           'InequalityConstraints', {g1, g2}, ...
                           'Bounds', salkinBounds, ...
                           'PopulationSize', 200, ...
                           'MaxGenerations', 200, ...
                           'PenaltyMultiplier', 1000.0);
    
    runResult = solver.solve();
    executionValues(run) = runResult.value;
    executionFevals(run) = runResult.fevalsSpent;
    
    if mod(run, 20) == 0
        fprintf('  > Completed %d/%d runs...\n', run, runsCount);
    end
end

% Compute statistics
avgFevals = mean(executionFevals);
bestValue = min(executionValues);
meanValue = mean(executionValues);
stdDeviation = std(executionValues);

% Display results
fprintf('\n=======================================================\n');
fprintf('          SALKIN PROBLEM BENCHMARK RESULTS             \n');
fprintf('=======================================================\n');
fprintf('Metric       |  Paper Table 2 Data  | MATLAB Run\n');
fprintf('-------------------------------------------------------\n');
fprintf('FEVALS       |  7244                | %d\n', round(avgFevals));
fprintf('BEST         |  320.0000            | %.4f\n', bestValue);
fprintf('MEAN         |  320.0000            | %.4f\n', meanValue);
fprintf('STD          |  0.0                 | %.2e\n', stdDeviation);
fprintf('=======================================================\n');
